File: jsimple-oauth/src/main/java/jsimple/oauth/extractors/TokenExtractor20Impl.java
/*
 * Copyright (c) 2012-2014, Microsoft Mobile
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 *
 * This code was adapted from the Scribe Java OAuth library
 * (https://github.com/fernandezpablo85/scribe-java), with modest changes made
 * for JUniversal.  Scribe itself is distributed under the MIT license.
 */

package jsimple.oauth.extractors;

import jsimple.oauth.model.Token;
import jsimple.oauth.utils.OAuthEncoder;
import jsimple.util.CharIterator;

/**
 * Default implementation of {@AccessTokenExtractor}. Conforms to OAuth 2.0
 */
public class TokenExtractor20Impl implements AccessTokenExtractor {
    private static final String EMPTY_SECRET = "";

    /**
     * {@inheritDoc}
     */
    public Token extract(String response) {
        // Original code regex:
        // private static final String TOKEN_REGEX = "access_token=([^&]+)";

        CharIterator tokenIterator = new CharIterator(response);
        tokenIterator.skipAheadPast("access_token=");
        StringBuilder tokenBuffer = new StringBuilder();
        while (!tokenIterator.isWhitespace() && tokenIterator.curr() != '&' && !tokenIterator.atEnd())
            tokenBuffer.append(tokenIterator.read());
        if (tokenBuffer.toString().isEmpty())
            throw new RuntimeException("oauth_token is empty string");

        String token = OAuthEncoder.decode(tokenBuffer.toString());
        return new Token(token, EMPTY_SECRET, response);
    }
}

File: jsimple-oauth/c#/jsimple/oauth/extractors/TokenExtractor20Impl.cs
using System;
using System.Text;

namespace jsimple.oauth.extractors
{

	using Token = jsimple.oauth.model.Token;
	using OAuthEncoder = jsimple.oauth.utils.OAuthEncoder;
	using CharIterator = jsimple.util.CharIterator;

	/// <summary>
	/// Default implementation of {@AccessTokenExtractor}. Conforms to OAuth 2.0
	/// </summary>
	public class TokenExtractor20Impl : AccessTokenExtractor
	{
		private const string EMPTY_SECRET = "";

		/// <summary>
		/// {@inheritDoc}
		/// </summary>
		public virtual Token extract(string response)
		{
			// Original code regex:
			// private static final String TOKEN_REGEX = "access_token=([^&]+)";

			CharIterator tokenIterator = new CharIterator(response);
			tokenIterator.skipAheadPast("access_token=");
			StringBuilder tokenBuffer = new StringBuilder();
			while (!tokenIterator.Whitespace && tokenIterator.curr() != '&' && !tokenIterator.atEnd())
				tokenBuffer.Append(tokenIterator.currAndAdvance());
			if (tokenBuffer.ToString().Length == 0)
				throw new Exception("oauth_token is empty string");

			string token = OAuthEncoder.decode(tokenBuffer.ToString());
			return new Token(token, EMPTY_SECRET, response);
		}
	}

}